File: micro/pipeline_3stage.c
// C equivalent of pipeline_3stage.hl.
// Three components connected by a subject-keyed router table — the
// same shape as bus_dispatch.c, just with two hops. Each publish is
// an FNV-hashed lookup + strcmp verify + indirect call (no queue,
// no memcpy like Hale's bus pays). Event and Filtered are both
// single-int payloads, so one handler signature covers both — same
// data as the Go version's typed maps.
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

static long long now_ns(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (long long)ts.tv_sec * 1000000000LL + (long long)ts.tv_nsec;
}

typedef void (*Handler)(void *ctx, long long value);

// Tiny string-keyed handler table (FNV-1a hash, linear probe,
// strcmp verify) — stands in for Go's map[string]func.
typedef struct {
    const char *subject;
    Handler fn;
    void *ctx;
} Route;

typedef struct {
    Route slots[4];
} Router;

static unsigned long long fnv1a(const char *s) {
    unsigned long long h = 1469598103934665603ULL;
    for (; *s; s++) {
        h ^= (unsigned char)*s;
        h *= 1099511628211ULL;
    }
    return h;
}

static void router_add(Router *r, const char *subject, Handler fn, void *ctx) {
    unsigned long long i = fnv1a(subject) & 3;
    while (r->slots[i].subject) i = (i + 1) & 3;
    r->slots[i] = (Route){subject, fn, ctx};
}

__attribute__((noinline)) static Route *router_lookup(Router *r, const char *subject) {
    unsigned long long i = fnv1a(subject) & 3;
    while (r->slots[i].subject) {
        if (strcmp(r->slots[i].subject, subject) == 0) return &r->slots[i];
        i = (i + 1) & 3;
    }
    return NULL;
}

typedef struct {
    long long count;
    long long sum;
} Sink;

__attribute__((noinline)) static void on_filtered(void *ctx, long long value) {
    Sink *s = ctx;
    s->count++;
    s->sum += value;
}

typedef struct {
    long long passed;
    Router *router;
} Filter;

__attribute__((noinline)) static void on_event(void *ctx, long long value) {
    Filter *f = ctx;
    if (value % 2 == 0) {
        f->passed++;
        Route *rt = router_lookup(f->router, "filtered"); // per-event, like the Go version
        if (!rt) exit(1);
        rt->fn(rt->ctx, value);
    }
}

typedef struct {
    long long count;
    Router *router;
} Source;

static void source_run(Source *s) {
    Route *emit = router_lookup(s->router, "event"); // hoisted, like the Go version
    if (!emit) exit(1);
    for (long long i = 0; i < s->count; i++) {
        emit->fn(emit->ctx, i);
    }
}

int main(void) {
    long long n = 50000;
    Sink sink = {0, 0};
    Router filtered_router = {0};
    router_add(&filtered_router, "filtered", on_filtered, &sink);
    Filter filter = {0, &filtered_router};
    Router event_router = {0};
    router_add(&event_router, "event", on_event, &filter);
    Source source = {n, &event_router};

    long long t0 = now_ns();
    source_run(&source);
    long long elapsed = now_ns() - t0;

    printf("n=%lld\n", n);
    printf("filter_passed=%lld\n", filter.passed);
    printf("sink_count=%lld\n", sink.count);
    printf("sink_sum=%lld\n", sink.sum);
    printf("elapsed_ns=%lld\n", elapsed);
    return 0;
}

File: micro/pipeline_3stage.rs
// Rust equivalent of pipeline_3stage.hl.
// Three components connected by a subject-keyed router map — the
// same shape as bus_dispatch.rs, just with two hops. Each publish is
// a map lookup + indirect (dyn Fn) call (no queue, no memcpy like
// Hale's bus pays). Cell counters stand in for Go's pointer-receiver
// mutation across the shared router graph.

use std::cell::Cell;
use std::collections::HashMap;
use std::time::Instant;

struct Event {
    value: i64,
}

struct Filtered {
    value: i64,
}

struct Sink {
    count: Cell<i64>,
    sum: Cell<i64>,
}

impl Sink {
    #[inline(never)]
    fn on_filtered(&self, f: Filtered) {
        self.count.set(self.count.get() + 1);
        self.sum.set(self.sum.get() + f.value);
    }
}

struct Filter<'a> {
    passed: Cell<i64>,
    router: HashMap<&'static str, Box<dyn Fn(Filtered) + 'a>>,
}

impl<'a> Filter<'a> {
    #[inline(never)]
    fn on_event(&self, e: Event) {
        if e.value % 2 == 0 {
            self.passed.set(self.passed.get() + 1);
            self.router["filtered"](Filtered { value: e.value });
        }
    }
}

struct Source<'a> {
    count: i64,
    router: HashMap<&'static str, Box<dyn Fn(Event) + 'a>>,
}

impl<'a> Source<'a> {
    fn run(&self) {
        let emit = &self.router["event"];
        for i in 0..self.count {
            emit(Event { value: i });
        }
    }
}

fn main() {
    let n: i64 = 50000;
    let sink = Sink {
        count: Cell::new(0),
        sum: Cell::new(0),
    };
    let mut filter_router: HashMap<&'static str, Box<dyn Fn(Filtered) + '_>> = HashMap::new();
    filter_router.insert("filtered", Box::new(|f| sink.on_filtered(f)));
    let filter = Filter {
        passed: Cell::new(0),
        router: filter_router,
    };
    let mut source_router: HashMap<&'static str, Box<dyn Fn(Event) + '_>> = HashMap::new();
    source_router.insert("event", Box::new(|e| filter.on_event(e)));
    let source = Source {
        count: n,
        router: source_router,
    };

    let t0 = Instant::now();
    source.run();
    let elapsed = t0.elapsed().as_nanos();

    println!("n={}", n);
    println!("filter_passed={}", filter.passed.get());
    println!("sink_count={}", sink.count.get());
    println!("sink_sum={}", sink.sum.get());
    println!("elapsed_ns={}", elapsed);
}
